File: engine/internal/api/benchmark.go
package api

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"

	"github.com/JennaLeigh311/Convergent-Routing-Analyzer/engine/internal/benchmark"
	"github.com/JennaLeigh311/Convergent-Routing-Analyzer/engine/internal/graph"
)

// endpointBenchmark and endpointBenchmarkStatus are the metric labels for the
// async benchmark endpoints.
const (
	endpointBenchmark       = "benchmark"
	endpointBenchmarkStatus = "benchmark_status"
)

// jobStatus is a benchmark job's lifecycle state.
type jobStatus string

const (
	statusRunning jobStatus = "running"
	statusDone    jobStatus = "done"
	statusFailed  jobStatus = "failed"
)

// Async benchmark resource bounds. The job store is long-lived and the sweep is
// CPU-heavy, so the endpoint is bounded on three axes so an unauthenticated
// caller cannot exhaust the process by varying the §R6 tuple:
//
//	maxRequestCount     — ceiling on the per-level OD count R, so one POST cannot
//	                      size an arbitrarily large (memory/CPU) sweep.
//	maxConcurrentSweeps — cap on in-flight async sweeps; a POST that would exceed
//	                      it gets a clean 503 instead of oversubscribing the box.
//	maxBenchmarkJobs    — cap on retained jobs; the store evicts the oldest
//	                      completed job (FIFO) rather than growing without bound.
const (
	maxRequestCount     = 100_000
	maxConcurrentSweeps = 4
	maxBenchmarkJobs    = 256
)

// benchmarkParams is the §R6 parameter tuple a /benchmark run is keyed by:
// (algorithm, α, β, capacity_scale, requestCount, seed). It is BOTH the request
// body and the cache key — two POSTs with the same tuple return the SAME job
// (the cached result) rather than re-running an expensive systemoptimal sweep.
//
// The `algorithm` field DISPATCHES between two modes, and every field flows into
// the result it selects — none are inert:
//
//   - algorithm == "all" (the default when omitted/empty) runs the canonical #91
//     six-router demand sweep, benchmark.RunSweep: it routes ALL six routers across
//     the FOUR v/c levels and SWEEPS cost.BPR.CapacityScale as its axis, so it reads
//     Seed and RequestCount and derives α/β/capacity-scale per level itself. In this
//     mode the request's α/β/CapacityScale do NOT alter the grid (the sweep owns
//     those); they remain part of the cache identity so the response echoes the
//     request faithfully. This path is unchanged and regression-safe.
//
//   - algorithm ∈ benchmark.RouterOrder (naive, reactive, incremental, msa,
//     systemoptimal, multipath) runs SINGLE-ALGORITHM mode, benchmark.RunSingle: it
//     builds ONE cost.BPR from the request's α/β/CapacityScale and routes that one
//     named router (plus the systemoptimal reference, for an honest PoA) at a single
//     level PINNED to the client's CapacityScale. Here α/β/CapacityScale actually
//     drive the cost function — two runs differing only in those params return
//     DIFFERENT metrics. This is the surface the Phase-6 parameter sliders (#104)
//     expose.
//
// Any other `algorithm` value is rejected at validate() time as a clean 400, never
// run. This keeps the endpoint a faithful composition over the existing harness
// (RunSweep / RunSingle) rather than a reimplementation; see docs/api.md.
type benchmarkParams struct {
	// Algorithm dispatches the run: "" / "all" is the full six-router sweep (the
	// harness default); one of benchmark.RouterOrder selects single-algorithm mode.
	Algorithm string `json:"algorithm"`
	// Alpha and Beta are the BPR coefficients (defaults 0.15 / 4 when zero).
	Alpha float64 `json:"alpha"`
	Beta  float64 `json:"beta"`
	// CapacityScale is the §R3 capacity knob (default 1.0 when zero).
	CapacityScale float64 `json:"capacity_scale"`
	// RequestCount is the per-level OD request count R (default DefaultODCount).
	RequestCount int `json:"request_count"`
	// Seed is the fixed RNG seed making the run reproducible (default 0).
	Seed int64 `json:"seed"`
}

// withDefaults returns a copy of p with zero fields filled from the harness
// defaults, so the cache key and the run agree on the effective tuple (two POSTs
// that differ only in an omitted-vs-explicit default hit the same cache entry).
func (p benchmarkParams) withDefaults() benchmarkParams {
	out := p
	if strings.TrimSpace(out.Algorithm) == "" {
		out.Algorithm = "all"
	}
	if out.Alpha == 0 {
		out.Alpha = 0.15
	}
	if out.Beta == 0 {
		out.Beta = 4
	}
	if out.CapacityScale == 0 {
		out.CapacityScale = 1.0
	}
	if out.RequestCount == 0 {
		out.RequestCount = benchmark.DefaultODCount
	}
	return out
}

// cacheKey is the canonical string form of the §R6 tuple, used to dedupe jobs. It
// is built from the defaulted params so the identity is stable regardless of which
// fields the client omitted.
func (p benchmarkParams) cacheKey() string {
	d := p.withDefaults()
	return fmt.Sprintf("algo=%s|a=%g|b=%g|cap=%g|n=%d|seed=%d",
		d.Algorithm, d.Alpha, d.Beta, d.CapacityScale, d.RequestCount, d.Seed)
}

// validate rejects out-of-contract params with a client-facing message, so a bad
// tuple is a clean 400 rather than a panic deep in the harness (NewBPR panics on
// a non-positive capacity scale; RunSweep would mis-size a negative count).
func (p benchmarkParams) validate() error {
	d := p.withDefaults()
	// algorithm must be "all" (the six-router sweep) or one of the six router names
	// (single-algorithm mode). Anything else is a client typo, rejected here as a 400
	// rather than dispatched into a run that has no such mode.
	if d.Algorithm != "all" && !benchmark.IsRouter(d.Algorithm) {
		return fmt.Errorf("algorithm must be \"all\" or one of %s", strings.Join(benchmark.RouterOrder, ", "))
	}
	if d.Alpha < 0 {
		return fmt.Errorf("alpha must be >= 0")
	}
	if d.Beta < 0 {
		return fmt.Errorf("beta must be >= 0")
	}
	if d.CapacityScale <= 0 {
		return fmt.Errorf("capacity_scale must be > 0")
	}
	if d.RequestCount < 0 {
		return fmt.Errorf("request_count must be >= 0")
	}
	if d.RequestCount > maxRequestCount {
		return fmt.Errorf("request_count must be <= %d", maxRequestCount)
	}
	return nil
}

// job is one benchmark run's tracked state. It is guarded by jobStore.mu; a
// handler never reads its fields without the lock.
type job struct {
	ID     string
	Params benchmarkParams
	Status jobStatus
	// Report is the harness output, populated when Status == done.
	Report *benchmark.Report
	// Err is the failure message when Status == failed (client-safe, no PII).
	Err string
}

// jobStore is the in-process registry of benchmark jobs: it maps a job id to its
// state and a §R6 cache key to the job that ran (or is running) that tuple, so a
// repeat POST returns the existing job instead of launching a duplicate
// systemoptimal sweep. It guards its maps with one mutex; the long-running sweep
// itself runs OUTSIDE the lock (the lock is held only to read/swap state), so a
// running benchmark never blocks a status poll.
//
// The store is BOUNDED: order tracks insertion order so that, once byID reaches
// maxJobs, the oldest COMPLETED job is evicted to make room — the maps cannot
// grow without bound as a caller varies the tuple. Running jobs are never evicted
// (their sweep is still writing back), but the concurrency cap keeps the running
// set tiny relative to maxJobs, so room is always reclaimable.
type jobStore struct {
	mu      sync.Mutex
	byID    map[string]*job
	byKey   map[string]string // cache key -> job id
	order   []string          // job ids in insertion order, for FIFO eviction
	maxJobs int
}

func newJobStore(maxJobs int) *jobStore {
	return &jobStore{
		byID:    make(map[string]*job),
		byKey:   make(map[string]string),
		maxJobs: maxJobs,
	}
}

// getOrCreate returns the existing job for params' cache key (cached or still
// running) and created=false, or registers a new running job and returns
// created=true. It is the dedupe point: the caller launches the sweep only when
// created is true, so the §R6 tuple maps to at most one run. When the store is at
// capacity it first evicts the oldest completed job (see evictOldestDoneLocked).
//
// It returns the job's id, status, and params BY VALUE (read under the lock) so
// the caller never reads the shared *job's fields off-lock — the async sweep
// mutates that job through complete() concurrently, so the handler must render
// its response from these copies, not from the live struct.
func (st *jobStore) getOrCreate(params benchmarkParams, id string) (gotID string, status jobStatus, gotParams benchmarkParams, created bool) {
	st.mu.Lock()
	defer st.mu.Unlock()
	key := params.cacheKey()
	if existingID, ok := st.byKey[key]; ok {
		j := st.byID[existingID]
		return j.ID, j.Status, j.Params, false
	}
	if st.maxJobs > 0 && len(st.byID) >= st.maxJobs {
		st.evictOldestDoneLocked()
	}
	j := &job{ID: id, Params: params.withDefaults(), Status: statusRunning}
	st.byID[id] = j
	st.byKey[key] = id
	st.order = append(st.order, id)
	return j.ID, j.Status, j.Params, true
}

// evictOldestDoneLocked removes the oldest job that is no longer running, freeing
// a store slot (and its retained Report). It is a no-op if every tracked job is
// still running — which the concurrency cap makes effectively impossible, since
// maxConcurrentSweeps is far below maxBenchmarkJobs. The caller must hold st.mu.
func (st *jobStore) evictOldestDoneLocked() {
	for i, id := range st.order {
		j, ok := st.byID[id]
		if ok && j.Status == statusRunning {
			continue
		}
		st.order = append(st.order[:i:i], st.order[i+1:]...)
		st.deleteLocked(id, j)
		return
	}
}

// deleteLocked drops a job from byID and, when byKey still points at it, byKey.
// It does not touch order (the caller owns that). The caller must hold st.mu.
func (st *jobStore) deleteLocked(id string, j *job) {
	delete(st.byID, id)
	if j != nil {
		if curID, ok := st.byKey[j.Params.cacheKey()]; ok && curID == id {
			delete(st.byKey, j.Params.cacheKey())
		}
	}
}

// remove drops a job entirely (byID, byKey, and order). It is the rollback path
// for a job that was created but could not be started (no sweep capacity), so a
// reserved-but-never-run job does not linger as a permanently "running" entry.
func (st *jobStore) remove(id string) {
	st.mu.Lock()
	defer st.mu.Unlock()
	j := st.byID[id]
	for i, oid := range st.order {
		if oid == id {
			st.order = append(st.order[:i:i], st.order[i+1:]...)
			break
		}
	}
	st.deleteLocked(id, j)
}

// complete records a finished sweep's report (or error) under the lock. A nil
// report with a non-empty errMsg marks the job failed; otherwise it is done. On
// failure it also drops the cache mapping for the tuple so the SAME params can be
// retried: a later POST starts a fresh run rather than returning this stuck
// failure forever. The failed job itself stays in byID (so a client already
// holding its id still polls the failure) until FIFO eviction reclaims it.
func (st *jobStore) complete(id string, report *benchmark.Report, errMsg string) {
	st.mu.Lock()
	defer st.mu.Unlock()
	j, ok := st.byID[id]
	if !ok {
		return
	}
	if errMsg != "" {
		j.Status = statusFailed
		j.Err = errMsg
		if curID, ok := st.byKey[j.Params.cacheKey()]; ok && curID == id {
			delete(st.byKey, j.Params.cacheKey())
		}
		return
	}
	j.Status = statusDone
	j.Report = report
}

// snapshot copies a job's current state under the lock so the handler renders a
// consistent view without holding the lock across JSON marshaling.
func (st *jobStore) snapshot(id string) (status jobStatus, params benchmarkParams, report *benchmark.Report, errMsg string, ok bool) {
	st.mu.Lock()
	defer st.mu.Unlock()
	j, found := st.byID[id]
	if !found {
		return "", benchmarkParams{}, nil, "", false
	}
	return j.Status, j.Params, j.Report, j.Err, true
}

// benchmarkStartResponse is the POST /benchmark body: the job id to poll plus the
// status, so a client knows immediately whether a fresh run started (running) or
// a cached identical run was returned (running/done/failed).
type benchmarkStartResponse struct {
	JobID  string          `json:"job_id"`
	Status jobStatus       `json:"status"`
	Params benchmarkParams `json:"params"`
}

// handleBenchmark serves POST /benchmark: it parses the §R6 tuple, returns a job
// id IMMEDIATELY, and runs the #91 harness async — the request NEVER blocks on a
// systemoptimal sweep. A repeat POST with the same tuple returns the existing job
// (the §R6 cache) instead of launching a duplicate run. The body is optional; an
// empty body runs the canonical harness defaults.
func (s *Server) handleBenchmark(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		s.writeError(w, endpointBenchmark, http.StatusMethodNotAllowed, "method not allowed: use POST")
		return
	}

	params, err := decodeBenchmarkParams(r.Body)
	if err != nil {
		s.writeError(w, endpointBenchmark, http.StatusBadRequest, "invalid request body: "+err.Error())
		return
	}
	if err := params.validate(); err != nil {
		s.writeError(w, endpointBenchmark, http.StatusBadRequest, err.Error())
		return
	}

	id, err := newJobID()
	if err != nil {
		s.logger.Error("generate job id failed", "endpoint", endpointBenchmark, "err", err)
		s.writeError(w, endpointBenchmark, http.StatusInternalServerError, "could not start benchmark")
		return
	}

	jobID, status, jobParams, created := s.jobs.getOrCreate(params, id)
	if created {
		// A fresh tuple needs a sweep slot. Reserve one without blocking; if the
		// box is already running maxConcurrentSweeps, roll the just-created job back
		// (so it does not linger as a never-completing "running" entry) and tell the
		// client to retry rather than oversubscribing CPU. The goroutine releases
		// the slot when the sweep finishes.
		select {
		case s.sweepSlots <- struct{}{}:
			s.runBenchmarkAsync(jobID, jobParams)
		default:
			s.jobs.remove(jobID)
			s.writeError(w, endpointBenchmark, http.StatusServiceUnavailable,
				"benchmark capacity reached, please retry shortly")
			return
		}
	}

	s.writeJSON(w, endpointBenchmark, http.StatusAccepted, benchmarkStartResponse{
		JobID:  jobID,
		Status: status,
		Params: jobParams,
	})
}

// sweepFunc is the seam runBenchmarkAsync runs the "all" (six-router sweep) path
// through: in production it is benchmark.RunSweep, and tests substitute a stub to
// exercise the failure, panic, and capacity paths without launching a real (slow)
// six-router sweep. It deliberately carries ONLY (seed, count) — the sweep owns
// α/β/capacity_scale (it sweeps capacity_scale as its axis), so widening this seam
// would only add params the sweep ignores. Single-algorithm mode does NOT go through
// this seam: it calls benchmark.RunSingle directly with the full param tuple (see
// runBenchmarkAsync), so the seam stays exactly the shape the existing tests stub.
type sweepFunc func(ctx context.Context, g graph.Graph, seed int64, count int) ([]benchmark.SweepCell, error)

// defaultSweep is the production "all" path: the canonical #91 six-router demand sweep.
var defaultSweep sweepFunc = benchmark.RunSweep

// runBenchmarkAsync launches the harness for the job in a goroutine and records
// the result on completion. It DISPATCHES on params.Algorithm: "all" runs the
// six-router sweep through the s.sweepFn seam (so the existing failure/panic/capacity
// tests stub exactly that path); any other algorithm — validated to be one of
// benchmark.RouterOrder — runs single-algorithm mode via benchmark.RunSingle with the
// full α/β/capacity_scale tuple, so those params actually drive the cost function.
//
// It uses context.Background() (NOT the request's context) so a client disconnecting
// after the immediate job-id response does not cancel an in-flight run — the job runs
// to completion and is cached for the next poll. A panic in the harness is recovered
// and recorded as a failed job rather than crashing the server. The caller has
// already reserved a sweep slot; this goroutine releases it on exit so the next
// queued POST can start.
func (s *Server) runBenchmarkAsync(id string, params benchmarkParams) {
	go func() {
		defer func() {
			<-s.sweepSlots
			if rec := recover(); rec != nil {
				s.logger.Error("benchmark job panicked", "endpoint", endpointBenchmark, "job_id", id, "panic", rec)
				s.jobs.complete(id, nil, "benchmark failed")
			}
		}()

		cells, err := s.runBenchmarkCells(params)
		if err != nil {
			s.logger.Error("benchmark run failed", "endpoint", endpointBenchmark, "job_id", id, "err", err)
			s.jobs.complete(id, nil, "benchmark failed")
			return
		}
		report := benchmark.BuildReport(params.Seed, params.RequestCount, cells)
		s.jobs.complete(id, &report, "")
		s.logger.Info("benchmark job complete", "endpoint", endpointBenchmark, "job_id", id,
			"algorithm", params.Algorithm, "cells", len(cells))
	}()
}

// runBenchmarkCells produces the comparison cells for a (validated) job's params,
// dispatching on Algorithm. "all" goes through the s.sweepFn test seam (the
// six-router sweep, regression-safe); a named router runs single-algorithm mode,
// building one BPR from the client's α/β/capacity_scale. params.Algorithm is assumed
// already defaulted+validated by the handler, so RunSingle only ever sees a name in
// benchmark.RouterOrder here.
func (s *Server) runBenchmarkCells(params benchmarkParams) ([]benchmark.SweepCell, error) {
	if params.Algorithm == "all" {
		return s.sweepFn(context.Background(), s.graph, params.Seed, params.RequestCount)
	}
	return benchmark.RunSingle(context.Background(), s.graph, params.Seed, params.RequestCount,
		params.Alpha, params.Beta, params.CapacityScale, params.Algorithm)
}

// benchmarkStatusResponse is the GET /benchmark/{id} body: the job's status, the
// §R6 tuple it ran, the cached report when done, and a client-safe error when
// failed. report is omitted while running.
type benchmarkStatusResponse struct {
	JobID  string            `json:"job_id"`
	Status jobStatus         `json:"status"`
	Params benchmarkParams   `json:"params"`
	Report *benchmark.Report `json:"report,omitempty"`
	Error  string            `json:"error,omitempty"`
}

// handleBenchmarkStatus serves GET /benchmark/{id}: it returns the job's
// status and, when done, the cached #91 metrics. An unknown id is a clean 404; a
// still-running job returns 200 with status "running" and no report.
func (s *Server) handleBenchmarkStatus(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		s.writeError(w, endpointBenchmarkStatus, http.StatusMethodNotAllowed, "method not allowed: use GET")
		return
	}

	id := strings.TrimPrefix(r.URL.Path, "/benchmark/")
	if id == "" || strings.Contains(id, "/") {
		s.writeError(w, endpointBenchmarkStatus, http.StatusBadRequest, "missing or malformed job id")
		return
	}

	status, params, report, errMsg, ok := s.jobs.snapshot(id)
	if !ok {
		s.writeError(w, endpointBenchmarkStatus, http.StatusNotFound, "no such benchmark job")
		return
	}

	s.writeJSON(w, endpointBenchmarkStatus, http.StatusOK, benchmarkStatusResponse{
		JobID:  id,
		Status: status,
		Params: params,
		Report: report,
		Error:  errMsg,
	})
}

// decodeBenchmarkParams decodes the optional JSON request body into a
// benchmarkParams. An empty body is the canonical defaults (not an error), so a
// bare `POST /benchmark` runs the standard harness. Unknown fields are rejected
// so a typo'd parameter fails loudly rather than silently running the default.
func decodeBenchmarkParams(body io.Reader) (benchmarkParams, error) {
	var params benchmarkParams
	if body == nil {
		return params, nil
	}
	dec := json.NewDecoder(io.LimitReader(body, maxBenchmarkBodyBytes))
	dec.DisallowUnknownFields()
	if err := dec.Decode(&params); err != nil {
		if err == io.EOF {
			// Empty body: run the defaults.
			return benchmarkParams{}, nil
		}
		return benchmarkParams{}, err
	}
	return params, nil
}

// maxBenchmarkBodyBytes bounds the request body so a runaway POST cannot exhaust
// memory; the tuple is tiny, so a small cap is generous.
const maxBenchmarkBodyBytes = 1 << 16

// newJobID returns a random, URL-safe job id. Crypto-random (not a counter) so a
// client cannot enumerate or guess other clients' job ids.
func newJobID() (string, error) {
	var buf [16]byte
	if _, err := rand.Read(buf[:]); err != nil {
		return "", err
	}
	return hex.EncodeToString(buf[:]), nil
}

File: web/src/lib/benchmark.ts
// POST /benchmark + GET /benchmark/{id} client for the before/after PoA panel (#102).
//
// The benchmark report is the SOURCE OF TRUTH for the money-shot numbers: the engine
// already computes the realized-time Price of Anarchy per demand level and the
// per-(router, level) mean/p95/total aggregates, so the frontend renders those
// verbatim rather than re-deriving a PoA from the live stream. POST /benchmark starts
// (or returns the cached) async sweep and 202s a job id; the report only arrives by
// polling GET /benchmark/{id} until status === "done" (a POST never carries the
// report itself), so runBenchmark always starts-then-polls.
//
// This client is deliberately small and self-contained so issue #104 (debounced
// async controls + result caching) can build its layer ON TOP of it without a
// rewrite: the param/response types mirror engine/internal/api/benchmark.go and the
// report types mirror engine/internal/benchmark, narrowed to the fields the panel
// consumes. The selectors (peakPoaLevel / selectLevelCells) are pure and tested.

import { restUrl } from "./engine";
import { ROUTER_ORDER, type Algo } from "./protocol";

/** A benchmark job's lifecycle state (engine: jobStatus). */
export type JobStatus = "running" | "done" | "failed";

// ---- §R6 parameter axis (issue #104) -----------------------------------------

/** The benchmark algorithm axis: "all" (six-router sweep) or one named router. */
export type BenchAlgo = "all" | Algo;

/** Selectable algorithm options for the #104 picker — engine order, "all" first. */
export const BENCH_ALGO_OPTIONS: readonly BenchAlgo[] = ["all", ...ROUTER_ORDER];

/**
 * The fully-defaulted §R6 tuple the #104 controls drive. It is identical in shape to
 * ResolvedBenchmarkParams (the engine's echoed params) but narrows `algorithm` to the
 * BenchAlgo union the picker offers. A full tuple is assignable to the all-optional
 * BenchmarkParams the HTTP client accepts, so it serializes verbatim as the POST body.
 */
export interface BenchmarkTuple extends ResolvedBenchmarkParams {
  algorithm: BenchAlgo;
}

/** The harness defaults, mirrored from benchmark.go withDefaults / DefaultODCount. */
export const DEFAULT_BENCH_PARAMS: BenchmarkTuple = {
  algorithm: "all",
  alpha: 0.15,
  beta: 4,
  capacity_scale: 1.0,
  request_count: 1000,
  seed: 0,
};

/**
 * Fill zero/empty fields from the harness defaults, mirroring the engine's
 * withDefaults() so the CLIENT cache key matches the SERVER's dedupe identity. The
 * contract treats a zero as "use the default" — e.g. alpha 0 is NOT a requestable BPR
 * coefficient (the engine maps it to 0.15) — so alpha 0 and alpha 0.15 are the SAME
 * tuple and must collapse to one cache entry here too.
 */
export function withBenchDefaults(p: BenchmarkTuple): BenchmarkTuple {
  return {
    algorithm: p.algorithm && String(p.algorithm).trim() !== "" ? p.algorithm : "all",
    alpha: p.alpha === 0 ? 0.15 : p.alpha,
    beta: p.beta === 0 ? 4 : p.beta,
    capacity_scale: p.capacity_scale === 0 ? 1.0 : p.capacity_scale,
    request_count: p.request_count === 0 ? 1000 : p.request_count,
    seed: p.seed,
  };
}

/**
 * Canonical string key for the §R6 tuple, the client-side cache key. Built from the
 * defaulted params so two tuples differing only in an omitted-vs-explicit default
 * collapse to one key — the same dedupe the server performs, so a repeat settle on an
 * identical tuple never re-fires a job (issue #104 acceptance).
 */
export function benchmarkKey(p: BenchmarkTuple): string {
  const d = withBenchDefaults(p);
  return `algo=${d.algorithm}|a=${d.alpha}|b=${d.beta}|cap=${d.capacity_scale}|n=${d.request_count}|seed=${d.seed}`;
}

// ---- classified errors -------------------------------------------------------

/** Classifies a benchmark failure so the UI can surface it specifically (#104). */
export type BenchErrorKind =
  | "bad_params" // 400: out-of-contract tuple
  | "capacity" // 503: box at max concurrent sweeps
  | "job_failed" // job ran but the harness failed (or finished without a report)
  | "not_found" // 404: polled an unknown/evicted job id
  | "network" // fetch threw (offline, CORS, …)
  | "unknown"; // any other non-OK status / timeout

/** A benchmark error carrying its classification and the originating HTTP status. */
export class BenchmarkError extends Error {
  readonly kind: BenchErrorKind;
  readonly status?: number;
  constructor(kind: BenchErrorKind, message: string, status?: number) {
    super(message);
    this.name = "BenchmarkError";
    this.kind = kind;
    this.status = status;
  }
}

/** Pull the engine's `{ error: string }` message off a non-OK response, best effort. */
async function errorMessage(res: Response, fallback: string): Promise<string> {
  try {
    const body = (await res.json()) as { error?: unknown };
    if (body && typeof body.error === "string" && body.error.trim() !== "") {
      return body.error;
    }
  } catch {
    // non-JSON / empty body — fall through to the fallback.
  }
  return fallback;
}

/**
 * The §R6 parameter tuple POST /benchmark is keyed by. All optional: an omitted
 * field takes the engine's default, and an empty/absent algorithm runs the canonical
 * six-router sweep ("all") — which is what the before/after panel wants (it needs the
 * naive vs systemoptimal per-level PoA the sweep produces).
 */
export interface BenchmarkParams {
  algorithm?: string;
  alpha?: number;
  beta?: number;
  capacity_scale?: number;
  request_count?: number;
  seed?: number;
}

/** The echoed, fully-defaulted tuple the engine ran (benchmarkParams). */
export interface ResolvedBenchmarkParams {
  algorithm: string;
  alpha: number;
  beta: number;
  capacity_scale: number;
  request_count: number;
  seed: number;
}

/** POST /benchmark body: the job id to poll plus the initial status. */
export interface StartResponse {
  job_id: string;
  status: JobStatus;
  params: ResolvedBenchmarkParams;
}

/** GET /benchmark/{id} body: status, the §R6 tuple, the report when done, error when failed. */
export interface StatusResponse {
  job_id: string;
  status: JobStatus;
  params: ResolvedBenchmarkParams;
  report?: BenchmarkReport;
  error?: string;
}

// ---- Report types (mirror engine/internal/benchmark) -------------------------

/** One (router, demand level) cell's realized-time metric bundle (benchmark.Result). */
export interface Result {
  router: string;
  demand_level: string;
  mean_realized_s: number;
  p95_realized_s: number;
  total_network_time_s: number;
  gini_vc: number;
  max_vc: number;
  iters: number;
  converged: boolean;
  gap: number;
}

/** One comparison-grid cell: the Result plus its cross-router PoA + sim columns. */
export interface SweepCell {
  result: Result;
  capacity_scale: number;
  target_vc: number;
  poa: number;
  sim_mean_realized_s: number;
  sim_p95_realized_s: number;
}

/** One demand level's headline naive-vs-systemoptimal Price of Anarchy. */
export interface LevelPoA {
  demand_level: string;
  target_vc: number;
  poa: number;
}

/** The headline improvement %, with the demand level it was measured at. */
export interface Improvement {
  demand_level: string;
  target_vc: number;
  best_router: string;
  percent_reduction: number;
  naive_total_s: number;
  best_total_s: number;
}

/** The full benchmark artifact (benchmark.Report). */
export interface BenchmarkReport {
  seed: number;
  od_count: number;
  total_demand_vph: number;
  cells: SweepCell[];
  poa_by_level: LevelPoA[];
  headline_improvement: Improvement;
}

// ---- HTTP ---------------------------------------------------------------------

/** POST /benchmark: start (or fetch the cached) sweep and return its job id.
 * Internal to runBenchmark; re-export if #104's controls ever need it directly. */
async function startBenchmark(
  params: BenchmarkParams = {},
  signal?: AbortSignal,
): Promise<StartResponse> {
  let res: Response;
  try {
    res = await fetch(restUrl("/benchmark"), {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(params),
      signal,
    });
  } catch (err) {
    if (signal?.aborted) throw err;
    const message = err instanceof Error ? err.message : "network error";
    throw new BenchmarkError("network", `POST /benchmark failed: ${message}`);
  }
  // Map the contract's failure statuses to a classified error so the caller can react
  // specifically (retry a 503, show a 400's reason) rather than swallowing them.
  if (res.status === 503) {
    throw new BenchmarkError("capacity", await errorMessage(res, "benchmark capacity reached, please retry shortly"), 503);
  }
  if (res.status === 400) {
    throw new BenchmarkError("bad_params", await errorMessage(res, "invalid benchmark parameters"), 400);
  }
  if (!res.ok) {
    throw new BenchmarkError("unknown", await errorMessage(res, `POST /benchmark failed: ${res.status} ${res.statusText}`), res.status);
  }
  return (await res.json()) as StartResponse;
}

/** GET /benchmark/{id}: poll one job's status (and report, once done).
 * Internal to runBenchmark; re-export if #104's controls ever need it directly. */
async function fetchBenchmarkStatus(
  jobId: string,
  signal?: AbortSignal,
): Promise<StatusResponse> {
  let res: Response;
  try {
    res = await fetch(restUrl(`/benchmark/${jobId}`), { signal });
  } catch (err) {
    if (signal?.aborted) throw err;
    const message = err instanceof Error ? err.message : "network error";
    throw new BenchmarkError("network", `GET /benchmark/${jobId} failed: ${message}`);
  }
  if (res.status === 404) {
    throw new BenchmarkError("not_found", await errorMessage(res, "no such benchmark job"), 404);
  }
  if (!res.ok) {
    throw new BenchmarkError("unknown", await errorMessage(res, `GET /benchmark/${jobId} failed: ${res.status} ${res.statusText}`), res.status);
  }
  return (await res.json()) as StatusResponse;
}

/** Options for runBenchmark's poll loop. */
export interface RunBenchmarkOptions {
  signal?: AbortSignal;
  /** Delay between status polls (ms). */
  pollIntervalMs?: number;
  /** Hard cap on poll attempts so a stuck job cannot loop forever. */
  maxPolls?: number;
  /** How many times to retry a 503 capacity rejection before surfacing it. */
  capacityRetries?: number;
  /** Base backoff between capacity retries (ms); grows linearly per attempt. */
  capacityBackoffMs?: number;
}

/** Reject if the signal is already aborted; otherwise resolve after ms (abortable). */
function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(new DOMException("aborted", "AbortError"));
      return;
    }
    const timer = setTimeout(() => {
      signal?.removeEventListener("abort", onAbort);
      resolve();
    }, ms);
    const onAbort = () => {
      clearTimeout(timer);
      reject(new DOMException("aborted", "AbortError"));
    };
    signal?.addEventListener("abort", onAbort, { once: true });
  });
}

/**
 * Start a benchmark and poll until it completes, resolving with the report. A POST
 * never carries the report (it only reserves/looks up the job), so this always polls
 * GET /benchmark/{id} at pollIntervalMs until status is "done" (resolve) or "failed"
 * (throw the engine's client-safe error). It is abortable via opts.signal and bounded
 * by maxPolls so a never-finishing job rejects rather than spinning forever. This is
 * the single entry point #104's debounced/cached controls reuse.
 */
export async function runBenchmark(
  params: BenchmarkParams = {},
  opts: RunBenchmarkOptions = {},
): Promise<BenchmarkReport> {
  const { signal, pollIntervalMs = 750, maxPolls = 240, capacityRetries = 3, capacityBackoffMs = 600 } = opts;

  // POST with bounded capacity retry/backoff: a 503 means the box is at its sweep cap
  // and asks the client to "retry shortly", so we back off briefly rather than failing
  // the run outright; any other error (or exhausting the retries) surfaces.
  let start: StartResponse | undefined;
  for (let attempt = 0; ; attempt++) {
    try {
      start = await startBenchmark(params, signal);
      break;
    } catch (err) {
      if (err instanceof BenchmarkError && err.kind === "capacity" && attempt < capacityRetries) {
        await sleep(capacityBackoffMs * (attempt + 1), signal);
        continue;
      }
      throw err;
    }
  }

  for (let i = 0; i < maxPolls; i++) {
    const status = await fetchBenchmarkStatus(start.job_id, signal);
    if (status.status === "done") {
      if (!status.report) {
        throw new BenchmarkError("job_failed", "benchmark reported done but carried no report");
      }
      return status.report;
    }
    if (status.status === "failed") {
      throw new BenchmarkError("job_failed", status.error || "benchmark failed");
    }
    await sleep(pollIntervalMs, signal);
  }
  throw new BenchmarkError("unknown", `benchmark did not finish within ${maxPolls} polls`);
}

// ---- Pure selectors over a report --------------------------------------------

/**
 * The PEAK per-level Price of Anarchy — the honest single money-shot number. PoA
 * peaks at moderate load and → 1 at both extremes (project-spec.md §5), so the peak
 * across the four sweep levels (carrying its demand level) is the figure to headline,
 * never a cherry-picked one. Returns null for a report with no per-level PoA.
 */
export function peakPoaLevel(report: BenchmarkReport): LevelPoA | null {
  let best: LevelPoA | null = null;
  for (const lvl of report.poa_by_level) {
    if (best === null || lvl.poa > best.poa) best = lvl;
  }
  return best;
}

/**
 * The cells at one demand level, indexed by router name. Used to pull naive (before)
 * vs reactive (after) vs systemoptimal (reference) for the mean/p95/total bar charts
 * at the headline level, straight from the report — no recomputation.
 */
export function selectLevelCells(
  report: BenchmarkReport,
  demandLevel: string,
): Record<string, SweepCell> {
  const out: Record<string, SweepCell> = {};
  for (const cell of report.cells) {
    if (cell.result.demand_level === demandLevel) {
      out[cell.result.router] = cell;
    }
  }
  return out;
}
